idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddExclExclCallFixFactories.kt

```diff
@@ -0,0 +1,121 @@
+        getFixForUnsafeCall(diagnostic.psi)
+    }
+
+    private fun KtAnalysisSession.getFixForUnsafeCall(psi: PsiElement): List<IntentionAction> {
+        val (target, hasImplicitReceiver) = when (psi) {
+            // `foo.bar` -> `foo!!.bar`
+            is KtDotQualifiedExpression -> psi.receiverExpression to false
+
+            // `foo[bar]` -> `foo!![bar]`
+            is KtArrayAccessExpression -> psi.arrayExpression to false
+
+            is KtCallableReferenceExpression -> psi.lhs.let { lhs ->
+                if (lhs != null) {
+                    // `foo::bar` -> `foo!!::bar`
+                    lhs to false
+                } else {
+                    // `::bar -> this!!::bar`
+                    psi to true
+                }
+            }
+
+            // `bar` -> `this!!.bar`
+            is KtNameReferenceExpression -> psi to true
+
+            // `bar()` -> `this!!.bar()`
+            is KtCallExpression -> psi to true
+
+            // `-foo` -> `-foo!!`
+            // NOTE: Unsafe unary operator call is reported as UNSAFE_CALL, _not_ UNSAFE_OPERATOR_CALL
+            is KtUnaryExpression -> psi.baseExpression to false
+
+            is KtBinaryExpression -> {
+                val receiver = if (KtPsiUtil.isInOrNotInOperation(psi)) {
+                    // `bar in foo` -> `bar in foo!!`
+                    psi.right
+                } else {
+                    // `foo + bar` -> `foo!! + bar` OR `foo infixFun bar` -> `foo!! infixFun bar`
+                    psi.left
+                }
+                receiver to false
+            }
+
+            // UNSAFE_INFIX_CALL/UNSAFE_OPERATOR_CALL on KtBinaryExpression is reported on the child KtOperationReferenceExpression
+            is KtOperationReferenceExpression -> return getFixForUnsafeCall(psi.parent)
+
+            else -> return emptyList()
+        }
+
+        return target.asFix(hasImplicitReceiver = hasImplicitReceiver)
+    }
+
+    // TODO: Move these factories to separate file when there are more *TYPE_MISMATCH fixes available.
+    // FE1.0 factory for TYPE_MISMATCH (QuickFixFactoryForTypeMismatchError) produces more than just AddExclExclFix.
+
+    val argumentTypeMismatchFactory = diagnosticFixFactory<KtFirDiagnostic.ArgumentTypeMismatch> { diagnostic ->
+        getFixesForTypeMismatch(diagnostic.psi, diagnostic.expectedType, diagnostic.actualType)
+    }
+
+    val returnTypeMismatchFactory = diagnosticFixFactory<KtFirDiagnostic.ReturnTypeMismatch> { diagnostic ->
+        getFixesForTypeMismatch(diagnostic.psi, diagnostic.expected, diagnostic.actual)
+    }
+
+    private fun KtAnalysisSession.getFixesForTypeMismatch(
+        psi: PsiElement,
+        expectedType: KtType,
+        actualType: KtType
+    ): List<IntentionAction> {
+        if (!expectedType.canBeNull && actualType.canBeNull) {
+            val nullableExpectedType = expectedType.makeNullable()
+            if (actualType isSubTypeOf nullableExpectedType) {
```

Comment:
Probably not in the scope of this PR. Should we store a boolean inside the `TYPE_MISMATCH` diagnostics to tell us whether the mismatch is sole due to nullability? I can see many quickfixes to benefit from this information. Otherwise, we have to do this subtype checking in a lot of places, which is expensive.

Reply:
Sounds good! But let's do it separately.